File: apple/BoundlessDriver/Sources/DriverShared/Onboarding/DriverOnboardingRouter.swift
import BoundlessKit
import RiderShared
import SwiftUI

/// Maps the current `OnboardingState` to the screen the Driver sees, wiring each affordance to the
/// view model. Like the Rider router, it owns only *view* state (the field text and the in-step
/// acknowledgements); the authoritative `OnboardingState` lives in the view model and is advanced
/// solely by the core (P4). The role-neutral steps reuse `RiderOnboardingScreens` verbatim; the
/// Driver-only screens come from `DriverOnboardingScreens`. Completion is **silent** — `.complete`
/// routes to the primary surface with no "all set" screen (voice-and-tone).
public struct DriverOnboardingRouter: View {
    @State private var viewModel: DriverOnboardingViewModel
    @State private var phone = ""
    @State private var code = ""
    // In-step view acknowledgements (the core state is unchanged until their Continue is tapped).
    @State private var recoveryCaptured = false
    @State private var permissionDeclinedAck = false
    @State private var autoUpdateEnabledAck = false

    public init(viewModel: DriverOnboardingViewModel) {
        _viewModel = State(initialValue: viewModel)
    }

    public var body: some View {
        Group {
            if viewModel.state == .complete {
                // Silent hand-off to the Driver's primary surface (home + Seat Toggle, a later spec).
                // Reuses the shared neutral placeholder — "no celebration of plumbing".
                PrimarySurfacePlaceholderView()
            } else {
                OnboardingScreenView(currentModel)
            }
        }
    }

    private var adminName: String? { viewModel.adminName }

    private var currentModel: OnboardingScreenModel {
        switch viewModel.state {
        case .freshInstall:
            return DriverOnboardingScreens.driverIntro(onContinue: viewModel.begin)

        case .phoneEntry:
            // A Driver routed back here by an invalidated session sees the re-auth variant
            // (`auth.signin_again`); a fresh install sees the plain phone entry. Both are the
            // core's `PhoneEntry` state — only the leading copy differs (AC15 Driver branch).
            if viewModel.reauthRequested {
                return DriverOnboardingScreens.reAuthPhoneEntry(text: $phone) {
                    Task { await viewModel.submitPhone(phone) }
                }
            }
            return RiderOnboardingScreens.phoneEntry(text: $phone) {
                Task { await viewModel.submitPhone(phone) }
            }

        case .phoneNotOnFile:
            return RiderOnboardingScreens.phoneNotOnFile(text: $phone, adminName: adminName) {
                Task { await viewModel.submitPhone(phone) }
            }

        case .deviceBinding:
            return RiderOnboardingScreens.deviceBinding(text: $code, adminName: adminName) {
                Task { await viewModel.submitCode(code) }
            }

        case .bindingFailed:
            return RiderOnboardingScreens.bindingFailed(text: $code, adminName: adminName) {
                Task { await viewModel.submitCode(code) }
            }

        case .permissions:
            // Driver-only: capture the one-time Recovery Code first, right after the device is bound
            // and the session issued (spec §C; ADR-0016 D3 / AC19), before the permissions ask. If no
            // code is available (degenerate shell case), skip it — never an empty-code screen, never
            // a block.
            if !recoveryCaptured, let recoveryCode = viewModel.recoveryCode {
                return DriverOnboardingScreens.recoveryCodeCapture(code: recoveryCode) {
                    recoveryCaptured = true
                }
            }
            if permissionDeclinedAck {
                return RiderOnboardingScreens.permissionsDeclined(adminName: adminName) {
                    Task { await viewModel.decideNotifications(allow: false) }
                }
            }
            return RiderOnboardingScreens.permissions(
                onAllow: { Task { await viewModel.decideNotifications(allow: true) } },
                onDecline: { permissionDeclinedAck = true }
            )

        case .autoUpdateStep:
            if autoUpdateEnabledAck {
                return RiderOnboardingScreens.autoUpdateEnabled(onContinue: viewModel.confirmAutoUpdate)
            }
            return RiderOnboardingScreens.autoUpdateStep(onContinue: { autoUpdateEnabledAck = true })

        case .belowMinVersion:
            // Calm degradation — no CTA, no form (O4 / AC8 / P10). A Driver can reach this from any
            // auth response just like a Rider.
            return RiderOnboardingScreens.calmHelp(adminName: adminName)

        case .needsReauthHelp:
            // Unreachable for a Driver (re-auth routes to `PhoneEntry`); mapped for exhaustiveness.
            return RiderOnboardingScreens.calmHelp(adminName: adminName)

        case .complete:
            // Unreachable: `.complete` is handled above by the placeholder. Kept for exhaustiveness.
            return RiderOnboardingScreens.calmHelp(adminName: adminName)
        }
    }
}

File: android/driver/app/src/main/kotlin/app/boundless/driver/onboarding/DriverOnboardingRouter.kt
package app.boundless.driver.onboarding

import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import app.boundless.driver.i18n.DriverStrings
import app.boundless.rider.onboarding.OnboardingScreenView
import app.boundless.rider.onboarding.PrimarySurfacePlaceholder
import app.boundless.rider.onboarding.RiderOnboardingScreens
import kotlinx.coroutines.launch
import uniffi.boundless_ffi_kotlin.OnboardingState

/**
 * Maps the current [OnboardingState] to the screen the Driver sees, wiring each affordance to the view
 * model. Like the Rider router, it owns only *view* state (the field text and the in-step
 * acknowledgements); the authoritative [OnboardingState] lives in the view model and is advanced solely
 * by the core (P4). The role-neutral steps reuse [RiderOnboardingScreens] verbatim (the shared kit);
 * the Driver-only screens come from [DriverOnboardingScreens]. Completion is **silent** — `COMPLETE`
 * routes to the primary surface with no "all set" screen (voice-and-tone). The Android twin of
 * `DriverShared.DriverOnboardingRouter`.
 *
 * The composition root (MainActivity, deferred to T14-shell) supplies the [strings] resolver and
 * constructs the [viewModel] with the real conformers (incl. the real `RecoveryCodeProviding`).
 */
@Composable
fun DriverOnboardingRouter(viewModel: DriverOnboardingViewModel, strings: DriverStrings) {
    // DriverStrings IS-A RiderStrings, so the shared factories take the same resolver (the Android
    // idiom for iOS reusing RiderShared's L10n alongside DriverL10n).
    val rider = remember(strings) { RiderOnboardingScreens(strings) }
    val driver = remember(strings) { DriverOnboardingScreens(strings) }
    val scope = rememberCoroutineScope()
    var phone by remember { mutableStateOf("") }
    var code by remember { mutableStateOf("") }
    // In-step view acknowledgements (the core state is unchanged until their Continue is tapped).
    var recoveryCaptured by remember { mutableStateOf(false) }
    var permissionDeclinedAck by remember { mutableStateOf(false) }
    var autoUpdateEnabledAck by remember { mutableStateOf(false) }

    val adminName = viewModel.adminName
    when (viewModel.state) {
        OnboardingState.COMPLETE ->
            // Silent hand-off to the Driver's primary surface (home + Seat Toggle, a later spec).
            PrimarySurfacePlaceholder()

        OnboardingState.FRESH_INSTALL ->
            OnboardingScreenView(driver.driverIntro(onContinue = viewModel::begin))

        OnboardingState.PHONE_ENTRY ->
            // A Driver routed back here by an invalidated session sees the re-auth variant
            // (`auth.signin_again`); a fresh install sees the plain phone entry. Both are the core's
            // `PhoneEntry` state — only the leading copy differs (AC15 Driver branch).
            if (viewModel.reauthRequested) {
                OnboardingScreenView(
                    driver.reAuthPhoneEntry { scope.launch { viewModel.submitPhone(phone) } },
                    fieldValue = phone,
                    onFieldValueChange = { phone = it },
                )
            } else {
                OnboardingScreenView(
                    rider.phoneEntry { scope.launch { viewModel.submitPhone(phone) } },
                    fieldValue = phone,
                    onFieldValueChange = { phone = it },
                )
            }

        OnboardingState.PHONE_NOT_ON_FILE ->
            OnboardingScreenView(
                rider.phoneNotOnFile(adminName) { scope.launch { viewModel.submitPhone(phone) } },
                fieldValue = phone,
                onFieldValueChange = { phone = it },
            )

        OnboardingState.DEVICE_BINDING ->
            OnboardingScreenView(
                rider.deviceBinding(adminName) { scope.launch { viewModel.submitCode(code) } },
                fieldValue = code,
                onFieldValueChange = { code = it },
            )

        OnboardingState.BINDING_FAILED ->
            OnboardingScreenView(
                rider.bindingFailed(adminName) { scope.launch { viewModel.submitCode(code) } },
                fieldValue = code,
                onFieldValueChange = { code = it },
            )

        OnboardingState.PERMISSIONS -> {
            // Driver-only: capture the one-time Recovery Code first, right after the device is bound
            // and the session issued (ADR-0016 D3 / AC19), before the permissions ask. If no code is
            // available (degenerate shell case), skip it — never an empty-code screen, never a block.
            val recoveryCode = viewModel.recoveryCode
            if (!recoveryCaptured && recoveryCode != null) {
                OnboardingScreenView(driver.recoveryCodeCapture(recoveryCode) { recoveryCaptured = true })
            } else if (permissionDeclinedAck) {
                OnboardingScreenView(
                    rider.permissionsDeclined(adminName) {
                        scope.launch { viewModel.decideNotifications(allow = false) }
                    },
                )
            } else {
                OnboardingScreenView(
                    rider.permissions(
                        onAllow = { scope.launch { viewModel.decideNotifications(allow = true) } },
                        onDecline = { permissionDeclinedAck = true },
                    ),
                )
            }
        }

        OnboardingState.AUTO_UPDATE_STEP ->
            if (autoUpdateEnabledAck) {
                OnboardingScreenView(rider.autoUpdateEnabled(onContinue = viewModel::confirmAutoUpdate))
            } else {
                OnboardingScreenView(rider.autoUpdateStep(onContinue = { autoUpdateEnabledAck = true }))
            }

        // Calm degradation (O4 / AC8 / P10). NEEDS_REAUTH_HELP is unreachable for a Driver (the core
        // routes Drivers to PHONE_ENTRY) but is mapped for `when` exhaustiveness.
        OnboardingState.BELOW_MIN_VERSION, OnboardingState.NEEDS_REAUTH_HELP ->
            OnboardingScreenView(rider.calmHelp(adminName))
    }
}
